src/BinSkim.Rules/PERules/BA6001.DisableIncrementalLinking.cs

```diff
@@ -0,0 +1,93 @@
+﻿/// <summary>
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Reflection.PortableExecutable;
+
+using Microsoft.CodeAnalysis.BinaryParsers;
+using Microsoft.CodeAnalysis.BinaryParsers.PortableExecutable;
+using Microsoft.CodeAnalysis.BinaryParsers.ProgramDatabase;
+using Microsoft.CodeAnalysis.IL.Sdk;
+using Microsoft.CodeAnalysis.Sarif;
+using Microsoft.CodeAnalysis.Sarif.Driver;
+
+namespace Microsoft.CodeAnalysis.IL.Rules
+{
+    [Export(typeof(ReportingDescriptor))]
+    [Export(typeof(Skimmer<BinaryAnalyzerContext>))]
+    public class DisableIncrementalLinking : WindowsBinaryAndPdbSkimmerBase
+    {
+        /// <summary>
+        /// BA6001
+        /// </summary>
+        public override string Id => RuleIds.DisableIncrementalLinking;
+
+        /// <summary>
+        /// Incremental linking support increases binary size and can reduce runtime performance. Fully optimized 
+        /// release builds should not specify incremental linking.
+        /// </summary>
+
+        public override MultiformatMessageString FullDescription =>
+            new MultiformatMessageString { Text = RuleResources.BA6001_DisableIncrementalLinking_Description };
+
+        protected override IEnumerable<string> MessageResourceNames => new string[] {
+                    nameof(RuleResources.BA6001_Pass),
+                    nameof(RuleResources.BA6001_Warning),
+                    nameof(RuleResources.BA6001_Pass_NonReleaseBuild)
+                };
+
+        private const string AnalyzerName = RuleIds.DisableIncrementalLinking + "." + nameof(DisableIncrementalLinking);
+
+        public override AnalysisApplicability CanAnalyzePE(PEBinary target, Sarif.PropertiesDictionary policy, out string reasonForNotAnalyzing)
+        {
+            PE portableExecutable = target.PE;
+            AnalysisApplicability result = AnalysisApplicability.NotApplicableToSpecifiedTarget;
+
+            reasonForNotAnalyzing = MetadataConditions.ImageIsILOnlyAssembly;
```

Comment:
>ImageIsILOnlyAssembly [](http://example.com/codeflow?start=55&length=21)

There's probably some analog here for .NET, to support edit-and-continue.

i.e., don't ship managed assemblies compiled for edit-and-continue.

I'm not sure what the risk is, relatively, from making this mistake in .NET vs. native binaries. also not clear what a reliable 'compiled for release' heuristic would be for that platform (but am guessing a fairly solid one could be contrived).

Anyway, just noting, if there's demand in future, we could consider updating this check.

Reply:
I have very limited knowledge of .NET so I am unsure whether a rule like this could apply to managed binaries.  It very well might!  I simply don't know enough.